CONTRIBUTING.md

```diff
@@ -125,10 +125,13 @@ versioning guidelines:
 
 # Runtime guarding
 
-Some high risk changes in Envoy are deemed worthy of runtime guarding. Instead of just replacing
+Some changes in Envoy are deemed worthy of runtime guarding. Instead of just replacing
 old code with new code, both code paths are supported for between one Envoy release (if it is
 guarded due to performance concerns) and a full deprecation cycle (if it is a high risk behavioral
-change).
+change). Generally as a community we try to guard both high risk changes (major
+refactors such as replacing Envoy's buffer implementation) and most user-visible
+non-config-guarded changes to HTTP processing (for example additions or changes to HTTP headers or
```

Comment:
I would probably s/HTTP/protocol and make the thing in parenthesis an example, but would be curious to hear what others think. At a high level I think we should be config guarding any user visible non-config-guarded change for any protocol extension that is not alpha. WDYT?

Reply:
Hm, I was trying to (possibly unsuccessfully) differentiate between changes like adding and removing headers, to changes like chunk coalescing or fixing infinite buffering.  do you think the memory fixes for chunk encoding and the gRPC infinite buffering fix should be guarded as well?  They're user-visible insofar as the user could be running a VM and consuming those chunks, and those logs.  I don't object to us getting that zealous but I wasn't arguing for it yet.

If you think it's time to guard all L7 changes I can retroactively guard the chunk PR - it's a  mild bummer to keep the literally buggy cruft around for 6 months but such is life for an increasingly widely used proxy. :-P

Reply:
> do you think the memory fixes for chunk encoding and the gRPC infinite buffering fix should be guarded as well?

No (unless we view them as high enough regression risk). I was mainly saying that we should apply the same rigor for non-HTTP protocols that are not alpha, such as Redis, etc.

> If you think it's time to guard all L7 changes I can retroactively guard the chunk PR - it's a mild bummer to keep the literally buggy cruft around for 6 months but such is life for an increasingly widely used proxy. :-P

IMO we can still apply maintainer discretion, but we should probably at least have the conversation for any such change?

Reply:
I'm happy to make the change you suggested - I'm trying to get clarity around "any user visible"
I tend towards "user visible changes likely to cause problems" where for HTTP header changes are often going to, things like changing hashing is definitely going to cause rollout pain, but chunking the body differently is very unlikely to be problematic.
Not sure how much we want to spell out and how much we want to say "maintainers believe will be problematic" 
I am also honestly curious about the gRPC thing.  I think if we limit log failure to flush size + buffer size it might be OK, but realistically we're going from never dropping to sometimes dropping and it's dicey :-P

Reply:
Also changing to protocol because I agree with your overall sentiment.  Might be good to have a !http example but I don't know enough about redis/kafka etc to suggest a likely to happen and user visible change.  Ideas?

Reply:
> Not sure how much we want to spell out and how much we want to say "maintainers believe will be problematic"

Yeah agreed. I think you can soften the language however you feel is appropriate.

> I am also honestly curious about the gRPC thing. I think if we limit log failure to flush size + buffer size it might be OK, but realistically we're going from never dropping to sometimes dropping and it's dicey :-P

Sorry I wasn't clear which fix you were talking about above. Yeah I agree it's dicey and could potentially benefit from a flag, though this is one of the cases that I don't feel strongly about.